File: crates/j-law-ruby/test/test_social_insurance.rb
# frozen_string_literal: true

require "minitest/autorun"
require "json"
require "date"
require "j_law_ruby"

class TestSocialInsurance < Minitest::Test
  FIXTURES = JSON.parse(File.read(File.join(__dir__, "../../../tests/fixtures/social_insurance.json")))

  FIXTURES["social_insurance"].each do |tc|
    define_method("test_#{tc['id']}") do
      inp = tc["input"]
      exp = tc["expected"]

      result = JLawRuby::SocialInsurance.calc_social_insurance(
        inp["standard_monthly_remuneration"],
        Date.parse(inp["date"]),
        inp["prefecture_code"],
        inp["is_care_insurance_applicable"]
      )

      assert_equal exp["health_related_amount"], result.health_related_amount
      assert_equal exp["pension_amount"], result.pension_amount
      assert_equal exp["total_amount"], result.total_amount
      assert_equal exp["care_insurance_applied"], result.care_insurance_applied?
    end
  end

  def test_invalid_standard_monthly_remuneration
    err = assert_raises(RuntimeError) do
      JLawRuby::SocialInsurance.calc_social_insurance(305_000, Date.new(2026, 3, 1), 13, false)
    end
    assert_match(/標準報酬月額/, err.message)
  end
end

File: crates/j-law-python/tests/test_social_insurance.py
"""社会保険料計算のテスト。"""

import datetime
import json
import pathlib

import pytest

from j_law_python.social_insurance import calc_social_insurance

_FIXTURE_PATH = (
    pathlib.Path(__file__).resolve().parents[3] / "tests" / "fixtures" / "social_insurance.json"
)
_FIXTURES = json.loads(_FIXTURE_PATH.read_text(encoding="utf-8"))


@pytest.mark.parametrize("case", _FIXTURES["social_insurance"], ids=lambda c: c["id"])
def test_social_insurance(case):
    inp = case["input"]
    exp = case["expected"]

    result = calc_social_insurance(
        inp["standard_monthly_remuneration"],
        datetime.date.fromisoformat(inp["date"]),
        inp["prefecture_code"],
        is_care_insurance_applicable=inp["is_care_insurance_applicable"],
    )

    assert result.health_related_amount == exp["health_related_amount"]
    assert result.pension_amount == exp["pension_amount"]
    assert result.total_amount == exp["total_amount"]
    assert result.care_insurance_applied is exp["care_insurance_applied"]


def test_social_insurance_invalid_standard_monthly_remuneration():
    with pytest.raises(ValueError, match="標準報酬月額"):
        calc_social_insurance(305_000, datetime.date(2026, 3, 1), 13)


def test_social_insurance_type_error_invalid_date():
    with pytest.raises(TypeError):
        calc_social_insurance(118_000, "2026-03-01", 13)
